Rendera/include/cube.hpp

```diff
@@ -0,0 +1,209 @@
+#ifndef CUBE_HPP
+#define CUBE_HPP
+
+#include "vec3.hpp"
+#include "ray.hpp"
+#include "material.hpp"
+
+#define INFINITY 1E10
+
+struct HitInfo
+{
+    bool is_hitting;
+    vec3 intersection_point;
+    Material material;
+    ray Ray;
+    int recursion_depth;
+
+    HitInfo(bool is_hitting, vec3 intersection_point, Material material, ray Ray, int recursion_depth) : is_hitting{is_hitting}, intersection_point{intersection_point}, material{material}, Ray{Ray}, recursion_depth{recursion_depth} {}
+};
+
+HitInfo max(HitInfo a, HitInfo b, vec3 o)
+{
+    if ((a.intersection_point - o).norm() > (b.intersection_point - o).norm())
+    {
+        return a;
+    }
+    else
+    {
+        return b;
+    }
+}
+
+HitInfo min(HitInfo a, HitInfo b, vec3 o)
+{
+
+    if ((a.intersection_point - o).norm() < (b.intersection_point - o).norm())
+    {
+        return a;
+    }
+    else
+    {
+        return b;
+    }
+}
+
+class Cube
+{
+private:
+    vec3 m_position;
+    vec3 m_a, m_b, m_c;
+    Material m_material;
+
+public:
+    Cube(vec3 position, vec3 a, vec3 b, vec3 c, Material mat) : m_position{position}, m_a{a}, m_b{b}, m_c{c}, m_material{mat} {};
+
+    ray normal(const vec3 &point)
+    {
+        vec3 p_pos, np_pos;
+        p_pos = point - m_position;
+        np_pos = point - m_a - m_b - m_c - m_position;
+        vec3 normal;
+        if (dot(p_pos, cross(m_a, m_b)) < 1e5)
+        {
+            normal = cross(m_a, m_b);
+            if (dot(m_c, normal) < 0)
+            {
+                normal = -normal;
+            }
+        }
+        else if (dot(p_pos, cross(m_a, m_c)) < 1e5)
+        {
+            normal = cross(m_a, m_c);
+            if (dot(m_b, normal) < 0)
+            {
+                normal = -normal;
+            }
+        }
+        else if (dot(p_pos, cross(m_b, m_c)) < 1e5)
+        {
+            normal = cross(m_b, m_c);
+            if (dot(m_a, normal) < 0)
+            {
+                normal = -normal;
+            }
+        }
+        else if (dot(np_pos, cross(m_a, m_b)) < 1e5)
+        {
+            normal = cross(m_a, m_b);
+            if (dot(-m_c, normal) < 0)
+            {
+                normal = -normal;
+            }
+        }
+        else if (dot(np_pos, cross(m_a, m_c)) < 1e5)
+        {
+            normal = cross(m_a, m_c);
+            if (dot(-m_b, normal) < 0)
+            {
+                normal = -normal;
+            }
+        }
+        else if (dot(np_pos, cross(m_b, m_c)) < 1e5)
+        {
+            normal = cross(m_b, m_c);
+            if (dot(-m_a, normal) < 0)
+            {
+                normal = -normal;
+            }
+        }
+        else
+        {
+            normal = vec3(0, 0, 0);
+        }
+        return ray(point, normal);
+    }
+
+    Material mat() const
```

Comment:
The `Material`, as well as the position of the object need not have getters, as this will be exposed via the contents of the `HitInfo` struct, which is returned by the `hit` function.